File: BackEnd/JavaScript/expressshop_prismaorm/src/auctions/auction.service.js
import {
  createAuction,
  createAuctionBid,
  deleteAuctionBid,
  findAuctionBidById,
  findAuctionBids,
  findAuctionById,
  findAuctions,
  findCategoryById,
  findSellerById,
  selectAuctionBid,
  updateAuction,
  updateAuctionBid,
} from "./auction.repository.js";
import { badRequest, forbidden, notFound } from "../utils/http-error.js";

const AUCTION_STATUSES = new Set(["OPEN", "CLOSED", "CANCELLED", "COMPLETED"]);

function isAuctionOpen(auction) {
  return auction.status === "OPEN" && auction.expiresAt > new Date();
}

function toAuctionSummary(auction) {
  return {
    id: auction.id,
    userId: auction.userId,
    userName: auction.user?.name ?? null,
    categoryId: auction.categoryId,
    categoryName: auction.category?.name ?? null,
    title: auction.title,
    budget: auction.budget,
    status: auction.status,
    bidCount: auction.bidCount,
    expiresAt: auction.expiresAt,
    createdAt: auction.createdAt,
    updatedAt: auction.updatedAt,
  };
}

function toBidDto(bid) {
  return {
    id: bid.id,
    auctionId: bid.auctionId,
    sellerId: bid.sellerId,
    sellerName: bid.seller?.name ?? null,
    price: bid.price,
    description: bid.description,
    deliveryDays: bid.deliveryDays,
    createdAt: bid.createdAt,
    updatedAt: bid.updatedAt,
  };
}

async function requireAuction(auctionId) {
  const auction = await findAuctionById(auctionId);
  if (!auction) {
    throw notFound("Auction not found");
  }
  return auction;
}

async function requireBid(auctionId, bidId) {
  const bid = await findAuctionBidById(auctionId, bidId);
  if (!bid) {
    throw notFound("Bid not found");
  }
  return bid;
}

async function resolveSellerFromUser(userId) {
  const seller = await findSellerById(userId);
  if (!seller) {
    throw badRequest("Seller profile not found");
  }
  return seller;
}

export async function getAuctions(query) {
  const page = Math.max(Number(query?.page ?? 1), 1);
  const limit = Math.min(Math.max(Number(query?.limit ?? 20), 1), 100);
  const status = query?.status ? String(query.status).toUpperCase() : null;
  const categoryId = query?.categoryId ? Number(query.categoryId) : null;

  if (status && !AUCTION_STATUSES.has(status)) {
    throw badRequest("status must be one of OPEN, CLOSED, CANCELLED, COMPLETED");
  }
  if (query?.categoryId !== undefined && Number.isNaN(categoryId)) {
    throw badRequest("categoryId must be a number");
  }

  const where = {
    ...(status ? { status } : {}),
    ...(categoryId ? { categoryId } : {}),
  };

  const [items, total] = await findAuctions(where, page, limit);
  return {
    items: items.map(toAuctionSummary),
    meta: {
      total,
      page,
      limit,
    },
  };
}

export async function getAuction(auctionId) {
  const auction = await requireAuction(auctionId);
  const bids = await findAuctionBids(auctionId);

  return {
    ...toAuctionSummary(auction),
    description: auction.description,
    specs: {},
    bids: bids.map(toBidDto),
  };
}

export async function createAuctionRequest(userId, payload) {
  if (!payload?.title) throw badRequest("title is required");
  if (!payload?.description) throw badRequest("description is required");
  if (!payload?.categoryId) throw badRequest("categoryId is required");

  const categoryId = Number(payload.categoryId);
  if (Number.isNaN(categoryId)) {
    throw badRequest("categoryId must be a number");
  }

  const category = await findCategoryById(categoryId);
  if (!category) {
    throw notFound("Category not found");
  }

  const created = await createAuction({
    userId: Number(userId),
    title: payload.title,
    description: payload.description,
    categoryId,
    budget: payload?.budget !== undefined && payload?.budget !== null ? Number(payload.budget) : null,
    status: "OPEN",
    bidCount: 0,
    expiresAt: new Date(Date.now() + 3 * 24 * 60 * 60 * 1000),
  });

  return getAuction(created.id);
}

export async function createBid(userId, auctionId, payload) {
  if (!payload?.price) throw badRequest("price is required");
  if (!payload?.deliveryDays) throw badRequest("deliveryDays is required");

  const price = Number(payload.price);
  const deliveryDays = Number(payload.deliveryDays);
  if (Number.isNaN(price) || Number.isNaN(deliveryDays)) {
    throw badRequest("price and deliveryDays must be numbers");
  }

  const seller = await resolveSellerFromUser(userId);
  const auction = await requireAuction(auctionId);
  if (!isAuctionOpen(auction)) {
    throw badRequest("Auction is not open");
  }

  const bid = await createAuctionBid(auctionId, {
    auctionId: Number(auctionId),
    sellerId: seller.id,
    price,
    description: payload?.description ?? null,
    deliveryDays,
  });

  return toBidDto({
    ...bid,
    seller,
  });
}

export async function selectBid(userId, auctionId, bidId) {
  const auction = await requireAuction(auctionId);
  if (auction.userId !== Number(userId)) {
    throw forbidden("Only the auction owner can select a bid");
  }
  if (!isAuctionOpen(auction)) {
    throw badRequest("Auction is not open");
  }

  const bid = await requireBid(auctionId, bidId);
  await selectAuctionBid(auctionId, bid.id);
  return { message: "낙찰을 선택했습니다." };
}

export async function cancelAuction(userId, auctionId) {
  const auction = await requireAuction(auctionId);
  if (auction.userId !== Number(userId)) {
    throw forbidden("Only the auction owner can cancel the auction");
  }
  if (!isAuctionOpen(auction)) {
    throw badRequest("Auction is not open");
  }

  await updateAuction(auctionId, { status: "CANCELLED" });
  return { message: "역경매가 취소되었습니다." };
}

export async function updateBid(userId, auctionId, bidId, payload) {
  const seller = await resolveSellerFromUser(userId);
  const auction = await requireAuction(auctionId);
  if (!isAuctionOpen(auction)) {
    throw badRequest("Auction is not open");
  }

  const bid = await requireBid(auctionId, bidId);
  if (bid.sellerId !== seller.id) {
    throw forbidden("Only the bid owner can update this bid");
  }

  const data = {};
  if (payload?.price !== undefined) data.price = Number(payload.price);
  if (payload?.description !== undefined) data.description = payload.description;
  if (payload?.deliveryDays !== undefined) data.deliveryDays = Number(payload.deliveryDays);

  if (Object.keys(data).length === 0) {
    throw badRequest("At least one field is required");
  }

  const updated = await updateAuctionBid(bid.id, data);
  return toBidDto({
    ...updated,
    seller,
  });
}

export async function deleteBid(userId, auctionId, bidId) {
  const seller = await resolveSellerFromUser(userId);
  const auction = await requireAuction(auctionId);
  if (!isAuctionOpen(auction)) {
    throw badRequest("Auction is not open");
  }

  const bid = await requireBid(auctionId, bidId);
  if (bid.sellerId !== seller.id) {
    throw forbidden("Only the bid owner can delete this bid");
  }

  await deleteAuctionBid(auctionId, bid.id);
  return { message: "입찰이 삭제되었습니다." };
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/auction/AuctionService.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.auction

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.PbShopException
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.StubResponse
import io.ktor.http.HttpStatusCode
import java.time.Instant

class AuctionService(
    private val repository: AuctionRepository,
) {
    fun createAuction(ownerId: Int, request: CreateAuctionRequest): StubResponse {
        validateAuctionRequest(request)
        request.categoryId?.let {
            if (!repository.categoryExists(it)) {
                throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "카테고리를 찾을 수 없습니다.")
            }
        }
        val created =
            repository.createAuction(
                ownerId = ownerId,
                input = NewAuction(request.title.trim(), request.description.trim(), request.categoryId, request.specs, request.budget, parseInstant(request.expiresAt)),
            )
        return StubResponse(status = HttpStatusCode.Created, data = detailPayload(created))
    }

    fun listAuctions(status: String?, categoryId: Int?, page: Int, limit: Int): StubResponse {
        val normalizedPage = if (page > 0) page else 1
        val normalizedLimit = limit.coerceIn(1, 100).takeIf { it > 0 } ?: 20
        val result = repository.listAuctions(parseStatus(status), categoryId, normalizedPage, normalizedLimit)
        return StubResponse(
            data = result.items.map(::summaryPayload),
            meta = mapOf("page" to normalizedPage, "limit" to normalizedLimit, "totalCount" to result.totalCount, "totalPages" to if (result.totalCount == 0) 0 else ((result.totalCount - 1) / normalizedLimit) + 1),
        )
    }

    fun detail(auctionId: Int): StubResponse = StubResponse(data = detailPayload(requireAuction(auctionId)))

    fun createBid(sellerId: Int, auctionId: Int, request: CreateAuctionBidRequest): StubResponse {
        requireOpenAuction(auctionId)
        if (!repository.sellerExists(sellerId)) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "판매자 권한이 필요합니다.")
        }
        if (request.price <= 0 || request.deliveryDays <= 0) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "유효한 입찰 정보를 입력해주세요.")
        }
        val bid = repository.createBid(sellerId, auctionId, NewAuctionBid(request.price, request.description?.trim()?.ifBlank { null }, request.deliveryDays))
        return StubResponse(status = HttpStatusCode.Created, data = bidPayload(bid))
    }

    fun selectBid(ownerId: Int, auctionId: Int, bidId: Int): StubResponse {
        val auction = requireOpenAuction(auctionId)
        if (auction.ownerId != ownerId) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "본인 경매에서만 낙찰을 선택할 수 있습니다.")
        }
        val bid = auction.bids.firstOrNull { it.id == bidId } ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "입찰을 찾을 수 없습니다.")
        if (bid.status != AuctionBidStatus.ACTIVE) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "활성 입찰만 선택할 수 있습니다.")
        }
        repository.selectBid(auctionId, bidId)
        return StubResponse(data = mapOf("message" to "낙찰이 선택되었습니다."))
    }

    fun cancelAuction(ownerId: Int, auctionId: Int): StubResponse {
        val auction = requireAuction(auctionId)
        if (auction.ownerId != ownerId) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "본인 경매만 취소할 수 있습니다.")
        }
        if (auction.status != AuctionStatus.OPEN) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "진행 중인 경매만 취소할 수 있습니다.")
        }
        repository.cancelAuction(auctionId)
        return StubResponse(data = mapOf("message" to "경매가 취소되었습니다."))
    }

    fun updateBid(sellerId: Int, auctionId: Int, bidId: Int, request: UpdateAuctionBidRequest): StubResponse {
        val auction = requireOpenAuction(auctionId)
        val bid = auction.bids.firstOrNull { it.id == bidId } ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "입찰을 찾을 수 없습니다.")
        if (bid.sellerId != sellerId) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "본인 입찰만 수정할 수 있습니다.")
        }
        if (request.price != null && request.price <= 0) throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "price는 0보다 커야 합니다.")
        if (request.deliveryDays != null && request.deliveryDays <= 0) throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "deliveryDays는 0보다 커야 합니다.")
        val updated = repository.updateBid(auctionId, bidId, AuctionBidUpdate(request.price, request.description?.trim(), request.deliveryDays))
            ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "입찰을 찾을 수 없습니다.")
        return StubResponse(data = bidPayload(updated))
    }

    fun deleteBid(sellerId: Int, auctionId: Int, bidId: Int): StubResponse {
        val auction = requireOpenAuction(auctionId)
        val bid = auction.bids.firstOrNull { it.id == bidId } ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "입찰을 찾을 수 없습니다.")
        if (bid.sellerId != sellerId) {
            throw PbShopException(HttpStatusCode.Forbidden, "FORBIDDEN", "본인 입찰만 취소할 수 있습니다.")
        }
        repository.deleteBid(auctionId, bidId)
        return StubResponse(data = mapOf("message" to "입찰이 취소되었습니다."))
    }

    private fun requireAuction(auctionId: Int): AuctionDetailRecord =
        repository.findAuctionById(auctionId) ?: throw PbShopException(HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND", "경매를 찾을 수 없습니다.")

    private fun requireOpenAuction(auctionId: Int): AuctionDetailRecord {
        val auction = requireAuction(auctionId)
        if (auction.status != AuctionStatus.OPEN) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "진행 중인 경매가 아닙니다.")
        }
        return auction
    }

    private fun validateAuctionRequest(request: CreateAuctionRequest) {
        if (request.title.trim().isBlank() || request.description.trim().isBlank()) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "title과 description이 필요합니다.")
        }
        if (request.budget != null && request.budget <= 0) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "budget은 0보다 커야 합니다.")
        }
        parseInstant(request.expiresAt)
    }

    private fun parseInstant(value: String?): Instant? =
        value?.trim()?.takeIf { it.isNotBlank() }?.let {
            runCatching { Instant.parse(it) }.getOrElse {
                throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "expiresAt 형식이 올바르지 않습니다.")
            }
        }

    private fun parseStatus(status: String?): AuctionStatus? =
        status?.trim()?.takeIf { it.isNotBlank() }?.uppercase()?.let {
            AuctionStatus.entries.firstOrNull { entry -> entry.name == it }
                ?: throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "지원하지 않는 auction status 입니다.")
        }

    private fun summaryPayload(record: AuctionSummaryRecord): Map<String, Any?> =
        mapOf(
            "id" to record.id,
            "ownerId" to record.ownerId,
            "title" to record.title,
            "categoryId" to record.categoryId,
            "budget" to record.budget,
            "status" to record.status.name,
            "bidCount" to record.bidCount,
            "createdAt" to record.createdAt.toString(),
        )

    private fun bidPayload(record: AuctionBidRecord): Map<String, Any?> =
        mapOf(
            "id" to record.id,
            "auctionId" to record.auctionId,
            "sellerId" to record.sellerId,
            "price" to record.price,
            "description" to record.description,
            "deliveryDays" to record.deliveryDays,
            "status" to record.status.name,
            "createdAt" to record.createdAt.toString(),
            "updatedAt" to record.updatedAt.toString(),
        )

    private fun detailPayload(record: AuctionDetailRecord): Map<String, Any?> =
        mapOf(
            "id" to record.id,
            "ownerId" to record.ownerId,
            "title" to record.title,
            "description" to record.description,
            "categoryId" to record.categoryId,
            "specs" to record.specsJson,
            "budget" to record.budget,
            "status" to record.status.name,
            "bidCount" to record.bidCount,
            "selectedBidId" to record.selectedBidId,
            "expiresAt" to record.expiresAt?.toString(),
            "bids" to record.bids.map(::bidPayload),
            "createdAt" to record.createdAt.toString(),
            "updatedAt" to record.updatedAt.toString(),
        )
}
